internal/ioutil/ioutil.go

```diff
@@ -101,21 +101,39 @@ type ioret[V any] struct {
 // If the deadline/context passes before the function finishes executing,
 // the zero value and the context error is returned.
 func WithDeadline[V any](ctx context.Context, timeout time.Duration, work func(ctx context.Context) (result V, err error)) (result V, err error) {
-	ctx, cancel := context.WithTimeout(ctx, timeout)
-	defer cancel()
+	return WithDeadlineAndReset[V](ctx, timeout, func(ctx context.Context, reset chan struct{}) (result V, err error) {
+		return work(ctx)
+	})
+}
+
+// WithDeadlineAndReset will execute a function with a deadline and return a value of a given type.
+// If the deadline/context passes before the function finishes executing, the zero value and the
+// context error is returned. The function can send a message to reset channel to delay deadline
+// triggering by another timeout
+func WithDeadlineAndReset[V any](ctx context.Context, timeout time.Duration, work func(ctx context.Context, reset chan struct{}) (result V, err error)) (result V, err error) {
+	t := time.NewTimer(timeout)
+	defer t.Stop()
 
+	reset := make(chan struct{})
```

Comment:
```suggestion
	reset := make(chan struct{}, 1)
```
`reset` needs to be a buffered to channel to avoid `work` function from being blocked due to an unbuffered channel with no receiver. The same reason why `c chan ioret[V]` is buffered.